File: test/extensions/language/expressions/cpp/LengthOfOperatorTest.cpp
#include "expressions/lengthof_operator/LengthOfFunctions.h"
#include "expressions/lengthof_operator/STR_CONSTANT.h"
#include "gtest/gtest.h"
#include "zserio/RebindAlloc.h"
#include "zserio/SerializeUtil.h"

namespace expressions
{
namespace lengthof_operator
{

using allocator_type = LengthOfFunctions::allocator_type;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;
using string_type = zserio::string<allocator_type>;

TEST(LengthOfOperatorTest, getLengthOfFixedArray)
{
    LengthOfFunctions lengthOfFunctions;
    const size_t fixedArrayLength = 10;
    vector_type<uint8_t> fixedArray(fixedArrayLength);
    lengthOfFunctions.setFixedArray(fixedArray);
    ASSERT_EQ(fixedArrayLength, lengthOfFunctions.funcGetLengthOfFixedArray());
}

TEST(LengthOfOperatorTest, getLengthOfVariableArray)
{
    LengthOfFunctions lengthOfFunctions;
    const size_t variableArrayLength = 11;
    vector_type<uint8_t> variableArray(variableArrayLength);
    lengthOfFunctions.setNumElements(static_cast<uint8_t>(variableArrayLength));
    lengthOfFunctions.setVariableArray(variableArray);
    ASSERT_EQ(variableArrayLength, lengthOfFunctions.funcGetLengthOfVariableArray());
}

TEST(LengthOfOperatorTest, getLengthOfStrConstant)
{
    LengthOfFunctions lengthOfFunctions;
    ASSERT_EQ(11, STR_CONSTANT.size()); // check that it's length in bytes (UTF-8)
    ASSERT_EQ(STR_CONSTANT.size(), lengthOfFunctions.funcGetLengthOfStrConstant());
}

TEST(LengthOfOperatorTest, getLengthOfLiteral)
{
    LengthOfFunctions lengthOfFunctions;
    ASSERT_EQ(10, zserio::makeStringView("€literal").size()); // check that it's length in bytes (UTF-8)
    ASSERT_EQ(zserio::makeStringView("€literal").size(), lengthOfFunctions.funcGetLengthOfLiteral());
}

TEST(LengthOfOperatorTest, literalLengthFieldDefault)
{
    LengthOfFunctions lengthOfFunctions;
    ASSERT_EQ(10, zserio::makeStringView("€literal").size()); // check that it's length in bytes (UTF-8)
    ASSERT_EQ(zserio::makeStringView("€literal").size(), lengthOfFunctions.getLiteralLengthField());
}

TEST(LengthOfOperatorTest, getLengthOfString)
{
    auto strField = string_type("€test");
    LengthOfFunctions lengthOfFunctions;
    lengthOfFunctions.setStrField(strField);
    ASSERT_EQ(7, strField.size()); // check that it's length in bytes (UTF-8)
    ASSERT_EQ(strField.size(), lengthOfFunctions.funcGetLengthOfString());
}

TEST(LengthOfOperatorTest, getLengthOfBytes)
{
    auto bytesField = vector_type<uint8_t>{{0x00, 0x01, 0x02}};
    LengthOfFunctions lengthOfFunctions;
    lengthOfFunctions.setBytesField(bytesField);
    ASSERT_EQ(bytesField.size(), lengthOfFunctions.funcGetLengthOfBytes());
}

TEST(LengthOfOperatorTest, getLengthOfFirstStrInArray)
{
    auto strArray = vector_type<string_type>{{string_type("€"), string_type("$")}};
    LengthOfFunctions lengthOfFunctions;
    lengthOfFunctions.setStrArray(strArray);
    ASSERT_EQ(3, strArray.at(0).size()); // check that it's length in bytes (UTF-8)
    ASSERT_EQ(strArray.at(0).size(), lengthOfFunctions.funcGetLengthOfFirstStrInArray());
}

TEST(LengthOfOperatorTest, getLengthOfFirstBytesInArray)
{
    auto bytesArray = vector_type<vector_type<uint8_t>>{{{{0x00, 0x01}}, {{}}}};
    LengthOfFunctions lengthOfFunctions;
    lengthOfFunctions.setBytesArray(bytesArray);
    ASSERT_EQ(bytesArray.at(0).size(), lengthOfFunctions.funcGetLengthOfFirstBytesInArray());
}

TEST(LengthOfOperatorTest, writeRead)
{
    LengthOfFunctions lengthOfFunctions;
    lengthOfFunctions.setFixedArray(
            vector_type<uint8_t>{{0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09}});
    lengthOfFunctions.setNumElements(3);
    lengthOfFunctions.setVariableArray(vector_type<uint8_t>{{0x03, 0x02, 0x01}});
    lengthOfFunctions.setStrField("longer than constant");
    lengthOfFunctions.setBytesField(vector_type<uint8_t>{{0x00, 0x01, 0x02}});
    lengthOfFunctions.setStrArray(vector_type<string_type>());
    lengthOfFunctions.setBytesArray(vector_type<vector_type<uint8_t>>());

    auto bitBuffer = zserio::serialize(lengthOfFunctions);
    auto readLengthOfFunctions = zserio::deserialize<LengthOfFunctions>(bitBuffer);
    ASSERT_EQ(lengthOfFunctions, readLengthOfFunctions);
}

} // namespace lengthof_operator
} // namespace expressions

File: test/extensions/language/expressions/python/LengthOfOperatorTest.py
import zserio

import Expressions


class LengthOfOperatorTest(Expressions.TestCase):
    def testGetLengthOfFixedArray(self):
        lengthOfFunctions = self.api.LengthOfFunctions()
        fixedArrayLength = 10
        lengthOfFunctions.fixed_array = list(range(fixedArrayLength))
        self.assertEqual(fixedArrayLength, lengthOfFunctions.get_length_of_fixed_array())

    def testGetLengthOfVariableArray(self):
        lengthOfFunctions = self.api.LengthOfFunctions()
        variableArrayLength = 11
        lengthOfFunctions.num_elements = variableArrayLength
        lengthOfFunctions.variable_array = list(range(variableArrayLength))
        self.assertEqual(variableArrayLength, lengthOfFunctions.get_length_of_variable_array())

    def testGetLengthOfStrContant(self):
        lengthOfFunctions = self.api.LengthOfFunctions()
        # check that it's length in bytes (UTF-8)
        self.assertEqual(11, zserio.builtin.lengthof_string("€constant"))
        self.assertEqual(
            zserio.builtin.lengthof_string(self.api.STR_CONSTANT),
            lengthOfFunctions.get_length_of_str_constant(),
        )

    def testGetLengthOfLiteral(self):
        lengthOfFunctions = self.api.LengthOfFunctions()
        # check that it's length in bytes (UTF-8)
        self.assertEqual(10, zserio.builtin.lengthof_string("€literal"))
        self.assertEqual(10, lengthOfFunctions.get_length_of_literal())

    def testLiteralLengthFieldDefault(self):
        lengthOfFunctions = self.api.LengthOfFunctions()
        # check that it's length in bytes (UTF-8)
        self.assertEqual(10, zserio.builtin.lengthof_string("€literal"))
        self.assertEqual(10, lengthOfFunctions.literal_length_field)

    def testGetLengthOfString(self):
        strField = "€test"
        lengthOfFunctions = self.api.LengthOfFunctions()
        lengthOfFunctions.str_field = strField
        # check that it's length in bytes (UTF-8)
        self.assertEqual(7, zserio.builtin.lengthof_string(strField))
        self.assertEqual(zserio.builtin.lengthof_string(strField), lengthOfFunctions.get_length_of_string())

    def testGetLengthOfBytes(self):
        bytesField = bytearray([0x00, 0x01, 0x02])
        lengthOfFunctions = self.api.LengthOfFunctions()
        lengthOfFunctions.bytes_field = bytesField
        self.assertEqual(len(bytesField), lengthOfFunctions.get_length_of_bytes())

    def testGetLengthOfFirstStrInArray(self):
        strArray = ["€", "$"]
        lengthOfFunctions = self.api.LengthOfFunctions()
        lengthOfFunctions.str_array = strArray
        # check that it's length in bytes (UTF-8)
        self.assertEqual(3, zserio.builtin.lengthof_string(strArray[0]))
        self.assertEqual(
            zserio.builtin.lengthof_string(strArray[0]),
            lengthOfFunctions.get_length_of_first_str_in_array(),
        )

    def testGetLengthOfFirstBytesInArray(self):
        bytesArray = [bytearray([0x00, 0x01]), bytearray([])]
        lengthOfFunctions = self.api.LengthOfFunctions()
        lengthOfFunctions.bytes_array = bytesArray
        self.assertEqual(len(bytesArray[0]), lengthOfFunctions.get_length_of_first_bytes_in_array())

    def testWriteRead(self):
        lengthOfFunctions = self.api.LengthOfFunctions()
        lengthOfFunctions.fixed_array = [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09]
        lengthOfFunctions.num_elements = 3
        lengthOfFunctions.variable_array = [0x03, 0x02, 0x01]
        lengthOfFunctions.str_field = "longer than constant"
        lengthOfFunctions.bytes_field = bytearray([0x00, 0x01, 0x02])
        lengthOfFunctions.str_array = []
        lengthOfFunctions.bytes_array = []

        bitBuffer = zserio.serialize(lengthOfFunctions)
        readLengthOfFunctions = zserio.deserialize(self.api.LengthOfFunctions, bitBuffer)
        self.assertEqual(lengthOfFunctions, readLengthOfFunctions)
